docs/development/new-cloud-provider.md

```diff
@@ -0,0 +1,56 @@
+# Adding Cloud Providers
+This document provides an overview of how to integrate a new cloud provider into Gardener. Each component that requires integration has a detailed description of how to integrate it and the steps required.
+
+## Cloud Components
+Gardener is composed of 2 or more Kubernetes clusters:
+
+* Shoot: These are one more "workload" clusters, the regular Kubernetes clusters you have seen. They provide places for your workloads to run.
+* Seed: This is the "management" cluster. It provides management service for one or more Shoot cluster.
+
+These two clusters can run in the same cloud provider, but they do not need to. For example, you could run your Seed in AWS, while having one shoot in Azure, two in Google, two in Alicloud, and three in Packet.
+
+The Seed cluster deploys and manages the Shoot clusters. Importantly, for this discussion, the `etcd` data store backing each Shoot runs as workloads inside the Seed. Thus, to use the above example, the clusters in Azure, Google, Alicloud and Packet will have their worker nodes and master nodes running in those clouds, but the `etcd` clusters backing them will run as separate [deployments](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) in the Seed Kubernetes cluster on AWS.
+
+This distinction becomes important when preparing to integrate a new cloud provider.
+
+## Gardener Cloud Integration
+Gardener integrates with cloud providers at the following key lifecycle elements:
+
+* Create/destroy/get/list machines for the Shoot
+* Create/destroy/get/list machines for the Seed
+* Backup/restore etcd for the Seed via writing files to and reading them from object storage
+
+Thus, the integrations you need for your cloud provider depend on whether you want to deploy Shoot clusters to the provider, Seed or both.
+
+* Shoot Only: machine lifecycle management
+* Seed: machine lifecycle management **and** etcd backup/restore
+
+## Gardener API
+In addition to the requirements to integrate with the cloud provider, you also need to enable the core Gardener app to receive, validate and process requests to use that cloud provider.
+
+* Expose the cloud provider to the consumers of the Gardener API, so it can be told to use that cloud provider as an option
+* Validate that API as requests come in
+* Extend Botanist to deploy machines to the cloud provider
+
+## Cloud Provider API Requirements
+In order for a cloud provider to integrate with Gardener, whether for Shoot or Seed clusters, the provider must have an API to perform machine lifecycle events, specifically:
+
+* Create a machine
+* Destroy a machine
+* Get information about a machine and its state
+* List machines
+
+As Gardener is written in [go](https://golang.org), the cloud provider should have a go SDK. However, if it has an API that is wrappable in go, e.g. a REST API, then you can use that to integrate.
```

Comment:
I don't get this block - why is this listed here? That's part of the machine controller manager, right? From Gardener PoV only a MCM docker image containing the cloud integration as well as a `MachineClass` must be delivered, there is no explicit machine handling done by Gardener.

Reply:
> I think the issue is that the docs for "gardener as a system" are in the same repo right here as "gardener as a component".

Yes. If you want to do that then I suggest to have a separate section/paragraph in the doc to not mix things up. Provide first information what needs to be done (ONLY from Gardener component PoV), and prior/later a system PoV.

WDYT?

Reply:
OK, that makes sense. I can clarify the language here.

How does that align with your comment [above](
https://github.com/gardener/gardener/pull/647#discussion_r247038326) about the Botanist interface? Doesn't it mean we need cloud provider integration in go (or written to wrap an API) for Gardener as well?

Reply:
No, it can be anything - Terraform, Go, whatever.

Reply:
I don't get it. 

* MCM: a single image, with integrations to AWS, GCE, Azure, Ali, Packet. If I want to add, e.g. DigitalOcean, I need to add it into the binary, which means go, does it not?
* Botanist: How would I implement the Botanist interface (a go construct) into Gardener if it is written in something other than go? Python or nodejs wouldn't work.

Reply:
OK, I got confused by what you mean. Sure, everything needs to be Golang as there is no way to have the providers externally yet. :)

Reply:
That is a good point. I will write something about in-tree vs out-of-tree pluggable here. Coming...

Reply:
Check it out now.